src/biogeochem/Fan2CTSMMod.F90

```diff
@@ -0,0 +1,1006 @@
+       end if
+       call relavu(unitn)
+    end if
+
+    call set_bcast_fanstream_pars(stream_year_first_fan,  stream_year_last_fan, &
+         model_year_align_fan, fan_mapalgo, stream_fldFileName_fan)
+
+    call shr_mpi_bcast(fan_to_bgc_crop, mpicom)
+    call shr_mpi_bcast(fan_to_bgc_veg, mpicom)
+    call shr_mpi_bcast(fract_spread_grass, mpicom)
+    call shr_mpi_bcast(nh4_ads_coef, mpicom)
+    
+    if (nh4_ads_coef < 0) then
+       call endrun(msg="ERROR invalid nh4_ads_coef")
+    end if
+    if (fract_spread_grass > 1 .or. fract_spread_grass < 0) then
+       call endrun(msg="ERROR invalid fract_spread_grass")
+    end if
+    
+  end subroutine fan_readnml
+
+  !************************************************************************************
+  
+  subroutine fan_eval(bounds, num_soilc, filter_soilc, &
+       atm2lnd_inst, wateratm2lndbulk_inst, &
+       cnveg_nitrogenflux_inst, &
+       soilbiogeochem_nitrogenflux_inst, &
+       soilbiogeochem_nitrogenstate_inst, &
+       waterstatebulk_inst, soilstate_inst, temperature_inst, &
+       waterfluxbulk_inst, frictionvel_inst)
+    use clm_time_manager, only: get_step_size, get_curr_date, get_curr_calday, get_nstep
+    use clm_varpar, only: max_patch_per_col
+    use LandunitType, only: lun
+    use shr_sys_mod, only : shr_sys_flush
+    use GridcellType, only: grc
+    use abortutils, only : endrun
+    use pftconMod, only : nc4_grass, nc3_nonarctic_grass
+    use landunit_varcon, only:  istsoil, istcrop
+    use clm_varcon, only : spval, ispval
+    use decompMod, only : bounds_type
+    use subgridAveMod, only: p2c
+
+    type(bounds_type)        , intent(in)    :: bounds  
+    integer                  , intent(in)    :: num_soilc       ! number of soil columns in filter
+    integer                  , intent(in)    :: filter_soilc(:) ! filter for soil columns
+    type(atm2lnd_type)       , intent(in)    :: atm2lnd_inst
+    type(wateratm2lndbulk_type), intent(in)  :: wateratm2lndbulk_inst
+    type(cnveg_nitrogenflux_type)          , intent(in) :: cnveg_nitrogenflux_inst
+    type(soilbiogeochem_nitrogenflux_type) , intent(inout) :: soilbiogeochem_nitrogenflux_inst
+    type(soilbiogeochem_nitrogenstate_type), intent(inout) :: soilbiogeochem_nitrogenstate_inst
+    type(waterstatebulk_type)              , intent(in)    :: waterstatebulk_inst
+    type(soilstate_type)                   , intent(in)    :: soilstate_inst
+    type(temperature_type)                 , intent(in) :: temperature_inst
+    type(waterfluxbulk_type)               , intent(in)    :: waterfluxbulk_inst
+    type(frictionvel_type)                 , intent(in) :: frictionvel_inst
+
+    ! Local variables
+    integer, parameter :: &
+         ! Use this many sub-steps. This improves numerical accuracy but is perhaps not
+         ! essential, because FAN includes an ad-hoc fixer for negative fluxes.
```

Comment:
The 'fixer' in FAN works so that if and N pool is becoming negative, all fluxes are scaled so as to keep it >=0. Personally I'd prefer this to be the last resort, so unless speed is an issue, I'd perhaps keep this.